src/plugins/saml-auth/actions.js

```diff
@@ -0,0 +1,173 @@
+  }
+}
+
+const parseLoginResponse = (response) => {
+
+  let data
+  try{ data = JSON.parse(response.data) }
+  catch(e) { return [undefined, e] }
+
+  const error = data && data.error
+  const parseError = data && data.parseError
+
+  switch(true) {
+    case !response.ok:
+      return [undefined, response.statusText]
+    case !!error:
+    case !!parseError:
+      return [data, JSON.stringify(data)]
+  }
+  return [data]
+}
+
+const authId = "SamlAuth"
+export const authenticateWithSamlToken = (schema, samlToken) => async ( { fn, samlAuthActions, authActions, errActions } ) => {
+  samlAuthActions.setSamlAuthState(SAML_AUTH_STATE_LOGGING_IN)
+
+  const method = schema.get("authMethod") || "post"
+  const authQuery = schema.get("authQuery")
+  const query = authQuery && authQuery.toJS()
+
+  const fetchUrl = appendQuery(schema.get("authTokenUrl"), query)
+
+  // 1. decode jwt
+  const [decoded, decodeErr] = decodeJWT(samlToken)
+  // guard error
+  if(decodeErr) {
+    errActions.newAuthErr({
+      authId,
+      level: "",
+      source: "auth",
+      message: `JWT Error: ${decodeErr.message}`
+    })
+    samlAuthActions.setSamlAuthState(SAML_AUTH_STATE_FAILED)
+    throw decodeErr
+  }
+
+  // 2. send request
+  const authBody = schema.get("authBody") || new Map()
+  const extraBody = authBody.toJS()
+  const body = JSON.stringify({ email: decoded.sub, saml_token: samlToken, ...extraBody })
+
+  const headers = {
+    "Accept":"application/json, text/plain, */*",
+    "Content-Type": "application/json"
+  }
+
+  let [response, fetchErr] = await (async function request() {
+    try {
+      return [await fn.fetch({
+        url: fetchUrl,
+        method: method || "post",
+        headers,
+        ...(method === "get" || method === "head" ? {} : {body})
+      }), undefined]
+    } catch(e) {
+      return [undefined, e]
+    }
+  })()
+  // guard error
+  if(fetchErr) {
+    const unauthorizedError = new Error(fetchErr)
+    unauthorizedError.message = fetchErr.response
+      ? fetchErr.response.body.message
+      : fetchErr.message
+
+    errActions.newAuthErr({
+      authId,
+      level: "",
+      source: "auth",
+      message: `Error logging in. ${unauthorizedError.message}`
+    })
+    samlAuthActions.setSamlAuthState(SAML_AUTH_STATE_FAILED)
+    throw unauthorizedError
+  }
+
+
+  // 3. parse response
+  const [data, parseErrorMessage] = parseLoginResponse(response)
+  if(parseErrorMessage) {
+    errActions.newAuthErr({
+      authId,
+      level: "",
+      source: "auth",
+      message: `Response Error: ${parseErrorMessage}`
+    })
+    samlAuthActions.setSamlAuthState(SAML_AUTH_STATE_FAILED)
+    throw new Error(parseErrorMessage)
+  }
+
+  // 4. set auth
+  authActions.authorize({
+    [authId]: {
+      name: authId,
+      email: decoded.sub,
+      token: data.token,
+      schema
+    }
+  })
+  samlAuthActions.saveSamlAuthEmail(decoded.sub)
+  samlAuthActions.setSamlAuthState(SAML_AUTH_STATE_LOGGED_IN)
+}
```

Comment:
There is no mixing of responsibility here. As mentioned,
```
EXTRACT & GENERALIZE authorizeOtpToken() & pass in diff body
```

The suggestion is not to mix both together but to extract out the token exchange part (e.g. `exchangeToken()`) because it is exactly the same including all error handling for token exchange, i.e.
- SAML auth - decode JWT & handle watever decoding error here, construct body `{email, SAMLToken}` & pass it to `exchangeToken()`
- OTP auth - construct body `{email, OTP}` & pass it to the same `exchangeToken()`

> 2. having different action at catch block, e.g. failed saml auth

This is not possible. The token exchange is aft a successful SAML auth. Even without this refactoring, the current impl `authenticateWithSamlToken = (authId, schema, samlToken) =>` it's a successful SAML auth expecting a samlToken to be passed in. In the event if SAML auth failed, this should not be called at all.

On why choosing the existing promise chaining approach, the reasons are mentioned in the comment ☝️ .